README
Variable-Strongly-Typed version 1.0.0

    This modules allow you to strongly type your variables.  Also known
    as the 'no fun' module - it can greatly enhance you code's quality
    and robustness.

    By enforcing types on some (or all) of your variables you will eliminate
    a large class of careless (& not so careless) errors.

    This could also aid an editor or code-browsing tools to verify code 
    correctness without having to execute the script.

    use Variable::Strongly::Typed;

    my $int             :TYPE('int');       # must have an 'int' value
    my $float           :TYPE('float');     # must have a 'float' value
    my $string          :TYPE('string');    # must not be a reference
    my $file            :TYPE('IO::File');  # must be an IO::File
    my @array_of_ints   :TYPE('int');       # Each slot must contain an int
    my %hash_of_floats  :TYPE('float');     # Each value must be a float

    my @array_of_rgb :TYPE(\&red_green_blue); # my enumerated type

    # ... and later ...

    $int = 23;          # All is well
    $int = 'howdy!';    # This line will croak with a good error message

    $float = 3.23;              # All is well, nothing to see here
    $float = new XML::Parser;   # croak!

    $array_of_ints[23] = 44;    # Groovy
    $array_of_ints[12] = 'yah'; # croak!

    $hash_of_floats{pi} = 3.14159;      # no problem
    $hash_of_floats{e}  = new IO::File; # croak!

    # Return 1 if this val is RED, BLUE, or GREEN
    #   0 otherwise
    sub red_green_blue {
        local $_ = shift;

        /\A RED \z/xms || /\A BLUE \z/xms || /\A GREEN \z/xms;
    }

    $array_of_my_very_own_types[23] = 99;       # croak!
    $array_of_my_very_own_types[2] = 'BLUE';    # OK!


INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


Alternatively, to install with Module::Build, you can use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

DEPENDENCIES

Class::Std::Utils
Attribute::Handlers

COPYRIGHT AND LICENCE

Copyright (C) 2005, Mark Ethan Trostler

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Variable-Strongly-Typed

1.0.0  Wed Aug 31 10:32:23 2005
       Initial release.



lib/Variable/Strongly/Typed/Hash.pm
package Variable::Strongly::Typed::Hash;

use version; $VERSION = qv('1.0.0');

use warnings;
use strict;
use Carp;

use base qw(Variable::Strongly::Typed);

{
    sub TIEHASH {
        my($class, $type) = @_;

        my $self = bless \my($anon_scalar), $class;
        $self->_init({}, $type);
    }

    sub STORE    { 
        my($self, $key, $value) = @_;
        my $hash = $self->_get_object;

        $self->_check_values($value);
        $hash->{$key} = $value;
    }

    sub FETCH    { 
        my($self, $key) = @_;
        my $hash = $self->_get_object;
        $hash->{$key};
    }

    sub FIRSTKEY { 
        my($self) = @_;
        my $hash = $self->_get_object;
        my $a = scalar keys %{$hash}; 
        each %{$hash};
    }

    sub NEXTKEY  { 
        my($self) = @_;
        my $hash = $self->_get_object;
        each %{$hash};
    }

    sub EXISTS   { 
        my($self, $key) = @_;
        my $hash = $self->_get_object;
        exists $hash->{$key}; 
    }

    sub DELETE   { 
        my($self, $key) = @_;
        my $hash = $self->_get_object;
        delete $hash->{$key};
    }   

    sub CLEAR    { 
        my($self) = @_;
        my $hash = $self->_get_object;
        %{$hash} = () 
    }

    sub SCALAR   { 
        my($self) = @_;
        my $hash = $self->_get_object;
        scalar %{$hash}; 
    }

}

1;

__END__

=head1 NAME

Variable::Strongly::Typed::Hash - Strongly typed hash


=head1 VERSION

This document describes Variable::Strongly::Typed::Hash version 1.0.0


=head1 SYNOPSIS

    This class is utilized by Variable::Strongly::Typed - you don't
    access this directly

 =head1 DESCRIPTION

    DO NOT USE THIS MODULE DIRECTLY!!
    It's used by Variable::Strongly::Typed to do its magic.

=head1 INTERFACE 
    
    None for you

=head1 DIAGNOSTICS

    None for you

=head1 CONFIGURATION AND ENVIRONMENT

    None

=head1 DEPENDENCIES

    Variable::Strongly::Typed

=head1 INCOMPATIBILITIES

    None

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-strongly-typed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Mark Ethan Trostler  C<< <mark@zzo.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005, Mark Ethan Trostler C<< <mark@zzo.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Variable/Strongly/Typed/Array.pm
package Variable::Strongly::Typed::Array;

use version; $VERSION = qv('1.0.0');

use warnings;
use strict;
use Carp;

use base qw(Variable::Strongly::Typed);

{
    sub TIEARRAY {
        my($class, $type) = @_;

        my $self = bless \my($anon_scalar), $class;
        $self->_init([], $type);
    }

    ### Bummer have to implement all the array functions...
    sub FETCHSIZE { 
        my $arr = shift->_get_object;
        return scalar @{$arr}; 
    }

    sub STORESIZE { 
        my $arr = shift->_get_object;
        $#{$arr} = $_[0]-1;
    }

    sub STORE     { 
        my ($self, $index, $value) = @_;
        my $arr = $self->_get_object;

        $self->_check_values($value);
        $arr->[$index] = $value;
    }

    # Note we don't currently check on the way out as
    #   Tie::Constrained does...
    sub FETCH     { 
        my $arr = shift->_get_object;
        $arr->[$_[0]];
    }

    sub CLEAR     { 
        my $arr = shift->_get_object;
        @{$arr} = ();
    }

    sub POP       { 
        my $arr = shift->_get_object;
        pop(@{$arr});
    }

    sub PUSH      { 
        my ($self, @values) = @_;
        my $arr = $self->_get_object;

        $self->_check_values(@values);
        push(@$arr, @values);
    }

    sub SHIFT     { 
        my $arr = shift->_get_object;
        shift(@{$arr});
    }

    sub UNSHIFT   { 
        my ($self, @values) = @_;
        my $arr = $self->_get_object;

        $self->_check_values(@values);
        unshift(@$arr, @values);
    }

    sub EXISTS    { 
        my $arr = shift->_get_object;
        exists $arr->[$_[0]];
    }

    sub DELETE    { 
        my $arr = shift->_get_object;
        delete $arr->[$_[0]];
    }
    
    sub SPLICE {
        my ($self) = shift;
        my $arr = $self->_get_object;

        my $sz  = $arr->FETCHSIZE;
        my $off = @_ ? shift : 0;
        $off   += $sz if $off < 0;
        my $len = @_ ? shift : $sz - $off;

        $self->_check_values(@_) if (@_);

        return splice(@$arr, $off, $len, @_);
    }

    sub EXTEND {
        my ($self) = shift;
        $self->STORESIZE(shift);
    }
}

1;

__END__

=head1 NAME

Variable::Strongly::Typed::Array - Strongly typed array


=head1 VERSION

This document describes Variable::Strongly::Typed::Array version 1.0.0


=head1 SYNOPSIS

    This class is utilized by Variable::Strongly::Typed - you don't
    access this directly

    my @array_of_ints   :TYPE('int');       # Each slot must contain an int
    my @array_of_rgb :TYPE(\&red_green_blue); # my enumerated type

    # ... and later ...

    $array_of_ints[23] = 44;    # Groovy
    $array_of_ints[12] = 'yah'; # croak!

    # Return 1 if this val is RED, BLUE, or GREEN
    #   0 otherwise
    sub red_green_blue {
        local $_ = shift;

        /\A RED \z/xms || /\A BLUE \z/xms || /\A GREEN \z/xms;
    }

    $array_of_my_very_own_types[23] = 99;       # croak!
    $array_of_my_very_own_types[2] = 'BLUE';    # OK!

=head1 DESCRIPTION

    DO NOT USE THIS MODULE DIRECTLY!!
    It's used by Variable::Strongly::Typed to do its magic.

=head1 INTERFACE 
    
    None for you

=head1 DIAGNOSTICS

    None for you

=head1 CONFIGURATION AND ENVIRONMENT

    None

=head1 DEPENDENCIES

    Variable::Strongly::Typed

=head1 INCOMPATIBILITIES

    None

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-strongly-typed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Mark Ethan Trostler  C<< <mark@zzo.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005, Mark Ethan Trostler C<< <mark@zzo.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Variable/Strongly/Typed/Scalar.pm
package Variable::Strongly::Typed::Scalar;

use version; $VERSION = qv('1.0.0');

use warnings;
use strict;
use Carp;

use base qw(Variable::Strongly::Typed);

{
    sub TIESCALAR {
        my($class, $type) = @_;

        my $self = bless \my($anon_scalar), $class;
        $self->_init(\my($another_anon_scalar), $type);
    }

    sub FETCH { 
        my $obj = shift->_get_object;
        return $$obj;
    }

    # If it's legit keep it
    sub STORE { 
        my ($self, $value) = @_;
        my $obj = $self->_get_object;

        $self->_check_values($value);
        $$obj = $value;
    }

  }

1;

__END__

=head1 NAME

Variable::Strongly::Typed::Scalar - Strongly typed scalar


=head1 VERSION

This document describes Variable::Strongly::Typed::Scalar version 1.0.0


=head1 SYNOPSIS

    This class is utilized by Variable::Strongly::Typed - you don't
    access this directly

 =head1 DESCRIPTION

    DO NOT USE THIS MODULE DIRECTLY!!
    It's used by Variable::Strongly::Typed to do its magic.

=head1 INTERFACE 
    
    None for you

=head1 DIAGNOSTICS

    None for you

=head1 CONFIGURATION AND ENVIRONMENT

    None

=head1 DEPENDENCIES

    Variable::Strongly::Typed

=head1 INCOMPATIBILITIES

    None

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-strongly-typed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Mark Ethan Trostler  C<< <mark@zzo.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005, Mark Ethan Trostler C<< <mark@zzo.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Variable/Strongly/Typed/Validators.pm
package Variable::Strongly::Typed::Validators;

use version; $VERSION = qv('1.0.0');

use warnings;
use strict;
use Carp;

our %conditions;

#
# Basic variable types
#
$conditions{int}    = '/\A\d+\z/xms';
$conditions{string} = '!ref';
$conditions{float}  =  '/( [-+]? (?:\d+\.?\d*|\.\d+) (?:[eE]\d+)? )/xms';
$conditions{bool}   =  $conditions{int};  # 0 already accounted for

1;

__END__

=head1 NAME

Variable::Strongly::Typed::Validators - Built-in type validators


=head1 VERSION

This document describes Variable::Strongly::Typed::Validators version 1.0.0


=head1 SYNOPSIS

    This class is utilized by Variable::Strongly::Typed - you don't
    access this directly.
    This module defines a '%conditions' hash contained expressions 
    that define the built-in types.
    If you want to create a new type just pass a CODE ref in your TYPE()
    attribute rather than modifying this hash.

 =head1 DESCRIPTION

    DO NOT USE THIS MODULE DIRECTLY!!
    It's used by Variable::Strongly::Typed to do its magic.

=head1 INTERFACE 
    
    None for you

=head1 DIAGNOSTICS

    None for you

=head1 CONFIGURATION AND ENVIRONMENT

    None

=head1 DEPENDENCIES

    Variable::Strongly::Typed

=head1 INCOMPATIBILITIES

    None

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-strongly-typed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Mark Ethan Trostler  C<< <mark@zzo.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005, Mark Ethan Trostler C<< <mark@zzo.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Variable/Strongly/Typed.pm
package Variable::Strongly::Typed;

use version; $VERSION = qv('1.0.0');

use warnings;
use strict;
use Carp;

# Tie any 'TYPE' attributes to me
use Attribute::Handlers;

# Our 3 friends for scalars & arrays & hashes
use Variable::Strongly::Typed::Scalar;
use Variable::Strongly::Typed::Array;
use Variable::Strongly::Typed::Hash;
use Variable::Strongly::Typed::Validators;

# Can't use Class::Std as it wants the 'ATTR' attribute
#   as does Attribute::Handlers...  Bummer
use Class::Std::Utils;
{
    # class variables
    my %actual_object_of;   # The actual object ref
    my %validator_sub_of;   # Validate subroutine
    my %type_of;            # Strong Type of this variable

    do {
        no warnings 'redefine';
        sub UNIVERSAL::TYPE : ATTR(SCALAR) {
            my ($package, $symbol, $referent, $attr, $data) = @_;
    
            tie $$referent, 'Variable::Strongly::Typed::Scalar', $data;
        }
    
        sub UNIVERSAL::TYPE : ATTR(ARRAY) {
            my ($package, $symbol, $referent, $attr, $data) = @_;
    
            tie @$referent, 'Variable::Strongly::Typed::Array', $data;
        }
    
        sub UNIVERSAL::TYPE : ATTR(HASH) {
            my ($package, $symbol, $referent, $attr, $data) = @_;
    
            tie %$referent, 'Variable::Strongly::Typed::Hash', $data;
        }
    };

    # Set up our class variables
    sub _init {
        my($self, $object, $type) = @_;
        my $ident = ident $self;

        $actual_object_of   {$ident} = $object;
        $validator_sub_of   {$ident} = $self->_make_validator_sub($type);
        $type_of            {$ident} = $type;

        return $self;
    }

    # Get actual object ref
    sub _get_object {
        my($self) = @_;
        return $actual_object_of{ident $self};
    }

    # Check if a value is okay...
    # IF this returns w/o croaking then all is okay...
    sub _check_values {
        my ($self, @values) = @_;

        my($valid_sub) = $validator_sub_of {ident $self};

        # if valid or failed sub says it's okay then ok
        #   otherwise undef
        # Note this works slightly differently then Tie::Constrained
        #   as we have no notion of an 'original' or 'default' value
        # 'failed' sub will croak if not valid...
        foreach my $val (@values) {

            # This'll croak if a value is not valid
            $valid_sub->($val) 
                || croak($self->_make_error_message($val));
        }

        # if we got this far all must be good
        return \@values;
    }

    sub _make_error_message {
        my($self, $val) = @_;

        # 'num', 'string', 'IO::File' or CODE ref if user defined
        my $valid_type = $type_of{ident $self};

        # Scalar, Array, or Hash
        my ($variable_type) = $self =~ /:: ([^:=]+) =/xms;
        $variable_type = lc $variable_type;

        # Type this variable is supposed to be
        my($strong_type) = $type_of{ident $self};

        # String-ify value
        if (ref $val) {
            $val = (ref $val) . ' reference';
        } else {
            $val = "'$val'";
        }

        my $msg;
        if (ref $valid_type eq 'CODE') {
            $msg = "Cannot assign $val to a user-validated"
                . " $variable_type variable!";
        } else {
            $msg = "Cannot assign $val to a" 
                . ($variable_type eq 'array' ? 'n' : '')
                . " $variable_type of type $strong_type!";
        }

        return $msg;
    }

    sub _make_validator_sub {
        my($self, $val) = @_;

        # User supplied??
        return $val if ($val && (ref($val) eq 'CODE'));

        my $condition;

        # Stuff like 'int', 'string', 'float' ... 'primitive' types
        $condition = $Variable::Strongly::Typed::Validators::conditions{$val};

        unless ($condition) {
            # It's either SCALAR, ARRAY, HASH the exact object type
            #   or an isa...  A reference type basically
            $condition = 
                'ref && ((ref eq ' . "'$val'" . ')'
                . ' || (ref ne "SCALAR" && ref ne "ARRAY"'
                .   ' && ref ne "HASH" && $_->isa(' . "'$val'" . ')))';
        }

        # Concoct validate sub
        my $sub = 
            eval 'sub { local $_ = shift; !$_ || (' . $condition . ') }';

        if ($@) {
            croak("Unable to create a Strongly Typed $val!!");
        }

        return $sub;
    }

    # Clean Up, Clean Up, Everybody's Doing Some, Clean Up
    #   -Walter (26 months)
    sub DESTROY {
        my($self) = @_;
        my $ident = ident $self;

        my $obj = $self->_get_object;
        undef  $obj;
        delete $actual_object_of{$ident};
        delete $validator_sub_of{$ident};
        delete $type_of         {$ident};
    }

}

1; # Magic true value required at end of module
__END__

=head1 NAME

Variable::Strongly::Typed - Let some variables be strongly typed


=head1 VERSION

This document describes Variable::Strongly::Typed version 1.0.0


=head1 SYNOPSIS

    use Variable::Strongly::Typed;

    my $int             :TYPE('int');       # must have an 'int' value
    my $float           :TYPE('float');     # must have a 'float' value
    my $string          :TYPE('string');    # must not be a reference
    my $file            :TYPE('IO::File');  # must be an IO::File
    my @array_of_ints   :TYPE('int');       # Each slot must contain an int
    my %hash_of_floats  :TYPE('float');     # Each value must be a float

    my @array_of_rgb :TYPE(\&red_green_blue); # my enumerated type

    # ... and later ...

    $int = 23;          # All is well
    $int = 'howdy!';    # This line will croak with a good error message

    $float = 3.23;              # All is well, nothing to see here
    $float = new XML::Parser;   # croak!

    $array_of_ints[23] = 44;    # Groovy
    $array_of_ints[12] = 'yah'; # croak!

    $hash_of_floats{pi} = 3.14159;      # no problem
    $hash_of_floats{e}  = new IO::File; # croak!

    # Return 1 if this val is RED, BLUE, or GREEN
    #   0 otherwise
    sub red_green_blue {
        local $_ = shift;

        /\A RED \z/xms || /\A BLUE \z/xms || /\A GREEN \z/xms;
    }

    $array_of_my_very_own_types[23] = 99;       # croak!
    $array_of_my_very_own_types[2] = 'BLUE';    # OK!

=head1 DESCRIPTION

    This modules allow you to strongly type your variables.  Also known
    as the 'no fun' module - it can greatly enhance you code's quality
    and robustness.

    By enforcing types on some (or all) of your variables you will eliminate
    a large class of careless (& not so careless) errors.

    This could also aid an editor or code-browsing tools to verify code 
    correctness without having to execute the script.

=head1 INTERFACE 

    Variable types are specified using the 'TYPE' attribute.  There are 
    currently 7 builtin types:

    'int'       => /\A \d+ \z/xms;
    'string'    => anything that's not a reference (!ref)
    'float'     => standard float regex
    'bool'      => same as 'int'
    'SCALAR'    => unblessed scalar ref
    'ARRAY'     => unblessed array ref
    'HASH'      => unblessed hash ref

    You can also specify a class name like 'XML::Parser' - note this is 
    NOT the same as 'HASH' which represents an unblessed HASH reference.

    Note when specifying 'XML::Parser' assigning 'XML::Parser' or any subclass of 
    'XML::Parser' will be valid - it does a $value->isa('XML::Parser')
    (or whatever the type you specified was) to test the value.

    If you try to assign an illegal value to your variable the assignment will
    croak with an informative error message.  Perhaps in the future there'll be
    an option to only carp...

    All builtin types allow 'undef' or any false value as a valid value.  Your
    own types (see section below) should probably do the same.

=head1 ROLLING YOUR OWN TYPES
    
    NOTE: You do NOT have to roll your own type to just specify a blessed
    object - see the section above!

    Specifying a CODE ref in your TYPE() attribute allows you to define your
    own types.  Your function will be call with 1 parameter - the value that's
    about to be assigned to any variable of that TYPE().

    Returning 1 (or any true value) will allow the assignment to happen.
    Return 0 (or any false value) will cause to assignment to croak.

    I'm sure you can dream up of all sorts of interesting types -
    enumerated types, all sorts of specific string types, bounded numbers,
    words only from a specific language are the sorts of things that pop
    into my mind.  Again I heartily expect you to surprise me with 
    the types you come up with.

    Note also you must allow for 'undef' as that is the initial value your
    variable will have!  Allowing for any false value may also be a good idea.

=head1 DIAGNOSTICS

    croak'ed error messages for built-in types look like:
        
      Cannot assign <value> to a [ scalar | hash | array ] of type <whatever>

    croak'ed error messages for user-supplied types look like:
        
      Cannot assign <value> to a user-validated  [ scalar | hash | array ] variable

=head1 CONFIGURATION AND ENVIRONMENT

Variable::Strongly::Typed requires no configuration files or environment variables.

=head1 DEPENDENCIES

    Attribute::Handlers handle the attribute stuff
    Class::Std::Utils handle the basic 'inside-out' module work
    Test::More for testing
    version for versioning

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-strongly-typed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Mark Ethan Trostler  C<< <mark@zzo.com> >>


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005, Mark Ethan Trostler C<< <mark@zzo.com> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
